File: DFS&BFS/python/2146.py
from collections import deque
import sys
input = sys.stdin.readline
sys.setrecursionlimit(10**6)
n = int(input())
graph = [list(map(int,input().split())) for _ in range(n)]
cnt = 2
moves = [(-1,0),(0,1),(1,0),(0,-1)]
land_list = [];
result = 10**18*9


def dfs(r,c,num):
    if graph[r][c] != 1 : return

    graph[r][c] = num
    for move in moves:
        nx,ny = move[0]+r , move[1] +c
        if nx>=n or ny>=n or nx<0 or ny<0: continue
        if graph[nx][ny] == 1: dfs(nx,ny,num)

def bfs(land_num):
    q = deque()
    distance_array = [[0 for _ in range(n)] for _ in range(n)]
    for i in range(n):
        for j in range(n):
            if graph[i][j] == land_num:
                q.append((i,j,0))
    while q:
        x,y,dis = q.popleft()
        for move in moves:
            nx,ny = x+move[0], y+move[1]
            if nx>=n or ny>=n or nx<0 or ny<0: continue
            if graph[nx][ny] == 0 and distance_array[nx][ny] == 0:
                distance_array[nx][ny] = dis+1
                q.append((nx,ny,dis+1))
            if graph[nx][ny] != land_num and graph[nx][ny] != 0:
                return dis




for i in range(n):
    for j in range(n):
        if graph[i][j] == 1:
            land_list.append(cnt)
            dfs(i,j,cnt)
            cnt+=1

while len(land_list) > 1:
    land_num = land_list.pop()
    result = min(result,bfs(land_num))
print(result)

File: DFS&BFS/javascript/2146.js
class Node {
  constructor(item = null){
    this.item = item;
    this.next = null;
  }
}
class Queue{
  constructor(){
    this.head = null;
    this.tail = null;
    this.size = 0;
  }
  enqueue(item){
    const newNode = new Node(item);
    if (this.size ===0 ) this.head = newNode;
    else this.tail.next = newNode;
    this.tail = newNode;
    this.size++;
  }
  dequeue(){
    if (this.size ===0) return -1;
    const popItem = this.head.item;
    this.head = this.head.next;
    if (this.head === null) this.tail = null;
    this.size--;
    return popItem;
  }
  isEmpty(){
    return this.size===0;
  }
}


const input = require('fs').readFileSync('/dev/stdin').toString().split('\n');
const n = Number(input[0]);
let graph = [];
const moves = [[-1,0],[0,1],[1,0],[0,-1]];
let move;
let nx,ny;
let landNum = 2;
let result = Infinity;

for (let i = 1 ; i<=n ; i++){
  graph.push(input[i].split(' ').map(Number));
}
const range_check = (x,y)=>{
  return ! ( x>=n || y>=n || x<0 || y<0);
}

const dfs = (r,c,landNum) =>{
  if (graph[r][c] !== 1 ) return
  graph[r][c] = landNum;
  for (move of moves){
    [nx,ny] = [r+move[0],c+move[1]];
    if (!range_check(nx,ny)) continue;
    if (graph[nx][ny] == 1) dfs(nx,ny,landNum);
  }
}
let visited = [];
for (let i = 0 ; i <n ; i++){
  visited.push(new Array(n));
}
const bfs = (landNum) => {
  for (let i = 0 ; i<n ; i++){
    for (let j = 0 ; j<n ; j++){
      visited[i][j] = false;
      if (graph[i][j] ===landNum){
        q.enqueue([i,j,0]);
      }
    }
  }
  
  let r,c,dis,nx,ny;
  while (!q.isEmpty()){
    [r,c,dis] = q.dequeue()
    for (move of moves){
      [nx,ny] = [r+move[0], c+move[1]];
      if (!range_check(nx,ny)) continue;
      if (graph[nx][ny] === 0 && !visited[nx][ny]){
        q.enqueue([nx,ny,dis+1]);
        visited[nx][ny] = true;
      }
      if (graph[nx][ny] !== 0 && graph[nx][ny] !== landNum){
        return dis;
      }
    }
  }
}
for (let i = 0 ; i<n ; i++){
  for (let j = 0 ; j<n ; j++){
    if (graph[i][j] == 1){
      dfs(i,j,landNum++);
    }
  }
}

let q = new Queue();
for (let i = 2 ; i<landNum-1 ; i++){
  while (! q.isEmpty()){
    q.dequeue();
  }
  result = Math.min(bfs(i),result);
}
console.log(result);